Analysis/Core/include/Analysis/RecoDecay.h

```diff
@@ -388,6 +401,70 @@ class RecoDecay
+    auto k = dotProd(flineXY, array{mom[0], mom[1]}) / Pt2(mom);
+    auto dx = flineXY[0] - k * (double)mom[0];
+    auto dy = flineXY[1] - k * (double)mom[1];
+    auto absImpPar = sqrtSumOfSquares(dx, dy);
+    auto fline = array{posSV[0] - posP[0], posSV[1] - posP[1], posSV[2] - posP[2]};
+    auto cross = crossProd(mom, fline);
+    return (cross[2] > 0. ? absImpPar : -1. * absImpPar);
+  }
+
+  /// Calculates the difference between measured and expected track impact parameter
+  /// normalized to its uncertainty
+  /// \param decLenXY decay lenght in {x, y} plane
+  /// \param errDecLenXY error on decay lenght in {x, y} plane
+  /// \param momMother {x, y, z} or {x, y} candidate momentum array
+  /// \param impParProng prong impact parameter
+  /// \param errImpParProng error on prong impact parameter
+  /// \param momProng {x, y, z} or {x, y} prong momentum array
+  /// \return normalized difference between expected and observed impact parameter
+  template <std::size_t N, typename T, typename U, typename V, typename Z>
+  static double NormImpParMeasMinusExpProng(T decLenXY, T errDecLenXY, const array<U, N>& momMother, V impParProng, V errImpParProng, const array<Z, N>& momProng)
+  {
+    // Ported from AliAODRecoDecayHF::Getd0MeasMinusExpProng
+    auto sinThetaP = ((double)momProng[0] * (double)momMother[1] - (double)momProng[1] * (double)momMother[0]) / (Pt(momProng) * Pt(momMother));
+    auto diff = impParProng - (double)decLenXY * sinThetaP;
+    auto errImpParExp = (double)errDecLenXY * sinThetaP;
+    auto errDiff = sqrtSumOfSquares(errImpParProng, errImpParExp);
+    return (errDiff > 0. ? diff / errDiff : 0.);
+  }
+
+  /// Calculates maximum normalized difference between measured and expected impact parameter of candidate prongs
+  /// \param posPV {x, y, z} or {x, y} position of primary vertex
+  /// \param posSV {x, y, z} or {x, y} position of secondary vertex
+  /// \param errDecLenXY error on decay lenght in {x, y} plane
+  /// \param momMother {x, y, z} or {x, y} candidate momentum array
+  /// \param arrImpPar array of prong impact parameters
+  /// \param arrErrImpPar array of errors on prong impact parameter (same order as arrImpPar)
+  /// \param momMom array of {x, y, z} or {x, y} prong momenta (same order as arrImpPar)
+  /// \return maximum normalized difference between expected and observed impact parameters
+  template <std::size_t N, std::size_t M, typename T, typename U, typename V, typename W, typename Z>
+  static double MaxNormImpPar(const T& posPV, const T& posSV, U errDecLenXY, const array<V, M>& momMother, const array<W, N>& arrImpPar, const array<W, N>& arrErrImpPar, const array<array<Z, M>, N>& arrMom)
+  {
+    auto decLenXY = distanceXY(posPV, posSV);
+    double maxNormIP{0.};
+    for (auto iProng = 0; iProng < N; ++iProng) {
+      auto prongNormIP = NormImpParMeasMinusExpProng(decLenXY, (double)errDecLenXY, momMother, arrImpPar[iProng], arrErrImpPar[iProng], arrMom[iProng]);
```

Comment:
Why is this needed? `(double)errDecLenXY`

Reply:
For the type template strategy that I adopted. Not needed anymore now